src/pgduckdb_duckdb.cpp

```diff
@@ -301,28 +323,30 @@ DuckDBManager::CreateConnection() {
 	return connection;
 }
 
-static bool transaction_handler_configured = false;
-bool started_duckdb_transaction = false;
-
 /* Returns the cached connection to the global DuckDB instance. */
 duckdb::Connection *
-DuckDBManager::GetConnection() {
+DuckDBManager::GetConnection(bool force_transaction) {
 	if (!pgduckdb::IsDuckdbExecutionAllowed()) {
 		elog(ERROR, "DuckDB execution is not allowed because you have not been granted the duckdb.postgres_role");
 	}
 
 	auto &instance = Get();
 	auto &context = *instance.connection->context;
 
-	if (!transaction_handler_configured) {
-		RegisterDuckdbXactCallback();
+	if (!context.transaction.HasActiveTransaction()) {
+		if (IsSubTransaction()) {
+			elog(ERROR, "SAVEPOINT is not supported in DuckDB");
+		}
+		if (IsInTransactionBlock() || force_transaction) {
+			/*
+			 * We only want to open a new DuckDB transaction if we're already
+			 * in a Postgres transaction block. Always opening a transaction
+			 * incurs a performance penalty when connecting to
```

Comment:
```suggestion
			 * incurs a performance penalty when connecting
```
or
```suggestion
			 * incurs a performance penalty when connecting to DuckDB
```

Reply:
added a lot more context here, I think I gave up writing this comment halfway through.